kolibri/auth/class_permissions_prototype_models.py

```diff
@@ -0,0 +1,170 @@
+#####################################################################################################################
+### This file is "realistic pseudocode" for the implementation of some of the class-based permissions features on ###
+### the models. Integration, and addition of tests, will take place following review of the general approach, and ###
+### discussion of next steps. Note that pieces of the models have been omitted here, as they were not needed for  ###
+### the purposes of illustration. Some suggestions for tweaks to the actual models will come as a separate PR.    ###
+#####################################################################################################################
+
+from .permissions import *
+
+ROLE_CHOICES = (
+    ("learner", _("Learner")),
+    ("coach", _("Coach")),
+    ("admin", _("Admin")),
+)
+
+class RoleQuerySet(models.QuerySet):
+    """
+    Queryset for the Role model with additional helper methods for efficiently querying by role-based/tree-based
+    relationships between users and other users or collections.
+    """
+
+    def __init__(self):
+
+        # retrieve the table names that will be used as context for building the query
+        self.table_names = {
+            "role_table": Role._meta.db_table,
+            "source_role": Role._meta.db_table, # alias to the Role table, to make later queries more readable
+            "collection_table": Collection._meta.db_table,
+        }
+
+        # list out the tables (along with aliases) that we'll need for use in the WHERE clause
+        self.tables = (
+            '"{collection_table}" AS "source_collection"',
+            '"{collection_table}" AS "target_collection"',
+            '"{role_table}" AS "target_role"',
+        )
+
+        # list out the primary role-based conditions for the WHERE clause
+        self.base_conditions = (
+            "{source_role}.collection_id = source_collection.id",
+            "{source_role}.type != 'learner'",
+            "target_collection.lft BETWEEN source_collection.lft AND source_collection.rght",
+        )
```

Comment:
Be sure to invoke the super method: https://github.com/django/django/blob/master/django/db/models/query.py#L167